File: TeamCode/src/main/java/com/info1robotics/bobot/tasks/AnalogTask.java
package com.info1robotics.bobot.tasks;

import android.drm.DrmStore;
import android.util.Pair;

import com.info1robotics.bobot.Common.GamepadEx;
import com.info1robotics.bobot.OpModes.templates.TeleOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class AnalogTask extends Task {
    public String button;
    public int gamepad;
    public double position = 0.0;

    public static class ActionPair {
        Type type;
        Task task;

        public ActionPair(Type type, Task task) {
            this.type = type;
            this.task = task;
        }
    }

    public enum Type {
        HOLD,
        RELEASE
    }

    public ArrayList<ActionPair> tasks = new ArrayList<>();

    public AnalogTask(String button, int gamepad, ActionPair ...children) {
        this.button = button;
        this.gamepad = gamepad;
        tasks.addAll(Arrays.asList(children));
    }

    @Override
    public void tick() {
        TeleOpMode teleOp = (TeleOpMode) context;
        tasks.forEach(pair -> {
            if (pair.task.isRunning()) pair.task.tick();
            GamepadEx gamepadEx = (gamepad == 1) ? teleOp.gamepadEx : teleOp.gamepadEx2;
            switch (pair.type) {
                case RELEASE:
                    if (position != gamepadEx.getAnalog(button) && gamepadEx.getAnalog(button) < 0.04) {
                        pair.task.start(context);
                    }
                    break;
                case HOLD:
                    if (gamepadEx.getAnalog(button) > 0.04) {
                        pair.task.start(context);
                    }
            }
            position = gamepadEx.getAnalog(button);
        });
    }
}

File: TeamCode/src/main/kotlin/com/info1robotics/bobot/tasks/AnalogTask.kt
package com.info1robotics.bobot.tasks

import com.info1robotics.bobot.Common.GamepadEx
import com.info1robotics.bobot.opmodes.TeleOpMode

/**
 * Listener for analog inputs.
 * Never ends, intended to be used with [AllTasks].
 * See [TaskBuilder.analog].
 * @param button The analog input to listen for.
 */
class AnalogTask(private val button: GamepadEx.Analog, private val gamepad: Int) : Task() {
    enum class Type {
        HOLD,
        RELEASE
    }
    var tasks = mutableMapOf<Type, SyncTasks>()
    var position = 0f

    override fun tick() {
        if (context !is TeleOpMode) {
            throw Exception("AnalogTask can only be used in TeleOp")
        }
        val teleOpMode = context as TeleOpMode
        val gamepadEx = if (gamepad == 1) teleOpMode.gamepadEx else teleOpMode.gamepadEx2
        tasks.forEach { (type, action) ->
            if (action.isRunning()) action.tick()
            when (type) {
                Type.HOLD -> {
                    if (gamepadEx.getAnalog(button) > 0.08) {
                        action.start(context)
                    }
                }
                Type.RELEASE -> {
                    if (position != gamepadEx.getAnalog(button) && gamepadEx.getAnalog(button) == 0f) {
                        action.start(context)
                    }
                }
            }
            position = gamepadEx.getAnalog(button)
        }
    }

    /**
     * Registers tasks to be synchronously executed when the button is executed with the given type.
     * @param type The action type. See [Type].
     * @param init Lambda to initialise the action. See [SyncTasks].
     */
    fun on(type: Type, init: SyncTasks.() -> Unit): SyncTasks {
        val task = SyncTasks()
        task.init()
        tasks[type] = task
        return task
    }
}